acceptance-tests/build.gradle

```diff
@@ -68,44 +58,46 @@ def allTestTasks =
             jenkinsVersion.default || currentJava.isCompatibleWith(jenkinsVersion.requiredJavaVersion)
         }
         .collect { jenkinsVersion ->
-            def jenkinsWarLocation =
-                new File("${project.gradle.gradleUserHomeDir}/jenkins-cache/${jenkinsVersion.version}/jenkins.war")
+            def downloadJenkinsTask =
+                tasks.register("download_jenkins_${jenkinsVersion.label}", Download) {
+                    src jenkinsVersion.downloadUrl
+                    dest file("${project.gradle.gradleUserHomeDir}/jenkins-cache/${jenkinsVersion.version}/jenkins.war")
+                    onlyIfModified true
+                    tempAndMove true
+                }
 
             def testTask =
                 jenkinsVersion.default
                     ? tasks.named("test")
                     : tasks.register("test_${jenkinsVersion.label}", Test)
 
             testTask.configure {
-                dependsOn copyTestDependencies
+                def jenkinsWar = downloadJenkinsTask
+                def gradlePlugin = configurations.gradlePlugin
+
+                inputs.files(jenkinsWar)
+                inputs.files(gradlePlugin)
 
                 onlyIf {
                     // Do not run on Windows as written here: https://github.com/jenkinsci/acceptance-test-harness/blob/master/docs/EXTERNAL.md
                     !OperatingSystem.current().isWindows()
                 }
 
-                doFirst {
-                    download.run {
-                        src jenkinsVersion.downloadUrl
-                        dest jenkinsWarLocation
-                        onlyIfModified true
-                        tempAndMove true
-                    }
-                }
-
-                // If executed on TeamCity, we need to set the Chromedriver path
-                if (gradle.ciTeamCityBuild) {
-                    systemProperty 'webdriver.chrome.driver', System.getenv('HOME') + '/.gradle/webdriver/chromedriver/chromedriver'
+                javaLauncher = javaToolchains.launcherFor {
+                    languageVersion = jenkinsVersion.javaVersion
                 }
 
-                environment([
-                    JENKINS_WAR: jenkinsWarLocation,
-                    LOCAL_JARS : testDependenciesDir.get().file("gradle.hpi"),
-                    BROWSER    : gradle.ciJenkinsBuild ? 'firefox-container' : 'chrome'
-                ])
+                doFirst {
+                    // If executed on TeamCity, we need to set the Chromedriver path
+                    if (gradle.ciTeamCityBuild) {
+                        systemProperty 'webdriver.chrome.driver', "${System.getenv('HOME')}/.gradle/webdriver/chromedriver/chromedriver"
```

Comment:
🤔 Why not use a command line argument provider for doing this, so it doesn't need to happen in a `doFirst`.

Reply:
There is no need to have it in the `doFirst` block. I can move it outside. Are there any downsides to having it in the `doFirst` block?

Reply:
Changing inputs in a `doFirst` block is a bad practice since it tries to circumvent Gradle's input/output detection and may cause undeclared inputs. We also want to deprecate being able to modify inputs in `doFirst` at some point.